// cookie-parser/index.js
'use strict';

function parseCookies(header) {
    const cookies = {};
    if (typeof header !== 'string' || header.length === 0) {
        return cookies;
    }
    const pairs = header.split(/;\s*/);
    for (const pair of pairs) {
        if (!pair) continue;
        const eqIndex = pair.indexOf('=');
        if (eqIndex === -1) {
            const name = decode(pair.trim());
            if (name) cookies[name] = '';
            continue;
        }
        const name = decode(pair.slice(0, eqIndex).trim());
        const value = decode(pair.slice(eqIndex + 1));
        if (!name) continue;
        cookies[name] = value;
    }
    return cookies;
}

function decode(value) {
    if (!value) return '';
    try {
        return decodeURIComponent(value.replace(/\+/g, ' '));
    } catch {
        return value;
    }
}

function cookieParser() {
    return function cookieParserMiddleware(req, _res, next) {
        if (req && typeof req === 'object') {
            const header = req.headers && req.headers.cookie;
            const parsed = parseCookies(header);
            if (!req.cookies || typeof req.cookies !== 'object') {
                req.cookies = {};
            }
            Object.assign(req.cookies, parsed);
            if (!req.signedCookies || typeof req.signedCookies !== 'object') {
                req.signedCookies = {};
            }
        }
        if (typeof next === 'function') next();
    };
}

module.exports = cookieParser;
module.exports.default = cookieParser;
module.exports.JSONCookies = function JSONCookies(obj) {
    return obj;
};
module.exports.signedCookie = function signedCookie(value) {
    return value;
};
